pkg/routeagent_driver/handlers/ovn/submarinerGWRouteHandler.go

```diff
@@ -0,0 +1,206 @@
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	delete(gatewayRouteHandler.remoteEndpoints, endpoint.Name)
+
+	if gatewayRouteHandler.isGateway {
+		if err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Delete(context.TODO(),
+			endpoint.Name, metav1.DeleteOptions{}); err != nil {
+			return errors.Wrapf(err, "error deleting gatewayRoute %q", endpoint.Name)
+		}
+	}
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) getNextHopOnK8sMgmtIntf(*submarinerv1.Endpoint) (*net.IP, error) {
+	link, err := netlink.LinkByName(OVNK8sMgmntIntfName)
+
+	if err != nil && !errors.Is(err, netlink.LinkNotFoundError{}) {
+		return nil, errors.Wrapf(err, "error retrieving link by name %q", OVNK8sMgmntIntfName)
+	}
+
+	currentRouteList, err := netlink.RouteList(link, syscall.AF_INET)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error retrieving routes on the link %s", OVNK8sMgmntIntfName)
+	}
+
+	for i := range currentRouteList {
+		logger.V(log.DEBUG).Infof("Processing route %v", currentRouteList[i])
+
+		if currentRouteList[i].Dst == nil || currentRouteList[i].Gw == nil {
+			continue
+		}
+
+		for _, subnet := range gatewayRouteHandler.config.ClusterCidr {
+			if currentRouteList[i].Dst.String() == subnet {
+				return &currentRouteList[i].Gw, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("could not find the route to %v via %q", gatewayRouteHandler.config.ClusterCidr, OVNK8sMgmntIntfName)
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) TransitionToNonGateway() error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	gatewayRouteHandler.isGateway = false
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) TransitionToGateway() error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	gatewayRouteHandler.isGateway = true
+
+	for _, endpoint := range gatewayRouteHandler.remoteEndpoints {
+		err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).DeleteCollection(context.TODO(),
+			metav1.DeleteOptions{}, metav1.ListOptions{
+				LabelSelector: "owner=submariner.io/ovn",
+			})
+		if err != nil {
+			return errors.Wrapf(err, "error deleting gatewayRoute %q", endpoint.Name)
+		}
+	}
+
+	for _, endpoint := range gatewayRouteHandler.remoteEndpoints {
+		if _, err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Create(context.TODO(),
+			gatewayRouteHandler.getGatewayRoute(endpoint), metav1.CreateOptions{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return errors.Wrapf(err, "error deleting gatewayRoute %q", endpoint.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) getGatewayRoute(endpoint *submarinerv1.Endpoint) *submarinerv1.GatewayRoute {
+	return &submarinerv1.GatewayRoute{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "GatewayRoute",
+			APIVersion: "submariner.io/v1",
+		},
```

Comment:
Why are you adding the type data here? This shouldn't be necessary.

Reply:
Without the type date , I am getting error.

Reply:
strange... what error and where?

Reply:
it was solved when we added the api to scheme.